### PR TITLE
Move UUID type to SPI module

### PR DESCRIPTION
**Pull request by ebyhr**
Prepare for #7097, #6663 and #5231

### CODE REVIEW

**Comment by electrum** [comment]
Type implementations shouldn't have to be in the SPI. Instead, connectors can retrieve the instance using `TypeManager`. However, not having the instance available statically causes difficulty for connectors like Cassandra that use an enum for types. Moving to use `TypeManager` could be significant refactoring. I'll leave this to @dain to make the call on this.

**Comment by electrum** [comment]
I'm fine with moving this to the SPI, but rather than making a decision about just the UUID type, I think it would be best to decide our stance on that for all built-in types.

**Comment by findepi** [comment]
> Prepare for #7097,

i hope Clickhouse is not dependent on this, see eg PostgreSQL connector.

Having said that, it seems UUID is really a standard type, and we are not getting a benefit from not having it in SPI.
Also, while the type implementation class may or may not be part of SPI, the type's internal representation *IS* part of SPI,
which is much more important aspect of type abstractions. Simply, we do not have -- and maybe we do not want to have -- 
real type abstractions for types that are pre-packaged. 
To me it seems the type abstractions provided by type system are really useful when a 3rd party plugin wants to use a type
provided by (same or different) 3rd party plugin, but that's not the case here.

File: core/trino-main/src/main/java/io/trino/type/TypeCoercion.java [hunk 1]
```diff
@@ -311,6 +311,7 @@ private TypeCompatibility typeCompatibilityForCovariantParametrizedType(Type fro
                 case StandardTypes.JSON:
                 case StandardTypes.INTERVAL_YEAR_TO_MONTH:
                 case StandardTypes.INTERVAL_DAY_TO_SECOND:
+                case StandardTypes.UUID:
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/type/TypeCoercion.java:314]: This seems unrelated to the first commit (StandardTypes.UUID existed before the change)
If this fixes something, can you please add a test?
Also, would it make sense to separate PR?

**Review by findepi — APPROVED** [review]
per https://github.com/trinodb/trino/pull/7801/#issuecomment-841060077

**Comment by findepi** [comment]
> causes difficulty for connectors like Cassandra that use an enum for types. Moving to use TypeManager could be significant refactoring

BTW i think this refactoring is unavoidable to support structural types properly.

**Comment by dain** [comment]
I don't care if it is in the SPI or not, but I would suggest that before moving to the SPI, you decide if the stack type should be `java.util.UUID` or should remain `Slice` (BTW we have supported custom stack types for a long time now).

**Comment by ebyhr** [comment]
Let me merge this PR.
